server/mqtt_ha.go

```diff
@@ -0,0 +1,114 @@
+package server
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/evcc-io/evcc/core/site"
+)
+
+type HADeviceDef struct {
+	Identifiers  string `json:"identifiers"`
+	Manufacturer string `json:"manufacturer"`
+	Name         string `json:"name"`
+}
+
+type HAEntityDef struct {
+	Device   HADeviceDef `json:"device"`
+	Name     string      `json:"name"`
+	UniqueId string      `json:"unique_id"`
+
+	StateTopic        string `json:"state_topic,omitempty"`
+	CommandTopic      string `json:"command_topic,omitempty"`
+	AvailabilityTopic string `json:"availability_topic,omitempty"`
+
+	Options []string `json:"options,omitempty"`
+
+	Min float64 `json:"min"`
+	Max float64 `json:"max"`
+
+	StateClass string `json:"state_class,omitempty"`
+}
+
+func (m *MQTT) haPublishBaseDeviceDef(site site.API, loadPoint *int, valueName string) HAEntityDef {
+	siteId := strings.ReplaceAll(strings.ToLower(site.Name()), " ", "_") // TODO find better site ID
+	uid := "evcc_" + siteId + "_"
+	if loadPoint != nil {
+		uid += fmt.Sprintf("lp%d_", *loadPoint+1)
+	}
+	uid += valueName
+
+	var name string
+	if loadPoint == nil {
+		name = fmt.Sprintf("EVCC %s %s", site.Name(), valueName)
+	} else {
+		name = fmt.Sprintf("EVCC %s Loadpoint %d %s", site.Name(), *loadPoint+1, valueName)
+	}
+
+	return HAEntityDef{
+		Device: HADeviceDef{
+			Identifiers:  fmt.Sprintf("evcc_%s", siteId),
+			Manufacturer: "EVCC",
+			Name:         site.Name(),
+		},
+		Name:              name,
+		UniqueId:          uid,
+		AvailabilityTopic: m.root + "/status",
+	}
+}
+
+func (m *MQTT) haPublishDiscoverSensors(site site.API, loadPoint *int, valueName, stateTopic string, val interface{}) {
+	if _, ok := m.haKnownSensors[stateTopic]; ok {
```

Comment:
wofür brauchst Du den check?

Reply:
Der Hauptgrund ist um zu verhinder das bei jeder Wert Änderung auch die HA discovery erneut gepusht wird (das ist nicht notwendig und produziert nur unnötig last).
Zum anderen hab ich damit auch verhindert das Sensoren für bereits definierte Aktoren (Select & Number) erstellt werden.

Reply:
> Der Hauptgrund ist um zu verhinder das bei jeder Wert Änderung auch die HA discovery erneut gepusht wird (das ist nicht notwendig und produziert nur unnötig last).

Das hab ich so nicht verstanden. Die ganzen Discovery Topics werden doch nur 1x definiert, genauso wie die normalen getter und setter?

Reply:
es gibt ja leider bei dem aktuelle Aufbau keine Liste aller existierenden Werte die per MQTT gesendet werden (zumindest habe ich da nix gefunden).
Deshalb werden die meißten sensoren erst in der Run loop des MQTT objekt definiert und hier kommen die Werte ja mehrfach. deshalb schau ich das ich die discovery topics nur einmal (beim ersten update) schreibe

Reply:
Ahhh. Du hast für jeden Wert einen Eintrag, nicht nur für die mit den Settern. Check.